marconi-core/src/Marconi/Core/Query.hs

```diff
@@ -112,21 +158,22 @@ allEvents :: EventsMatchingQuery event
 allEvents = EventsMatchingQuery Just
 
 -- | The result of an @EventMatchingQuery@
-type instance Result (EventsMatchingQuery event) = [Timed (Point event) event]
+type instance Result (EventsMatchingQuery event) = [Stability (Timed (Point event) event)]
 
 instance
   (MonadError (QueryError (EventsMatchingQuery event)) m)
   => Queryable m event (EventsMatchingQuery event) ListIndexer
   where
-  query p q ix = do
-    let isBefore p' e = p' >= e ^. point
-    let result =
-          mapMaybe (traverse $ predicate q) $
-            ix ^.. events . Lens.folded . Lens.filtered (isBefore p)
-
-    aheadOfSync <- isAheadOfSync p ix
-    when aheadOfSync $ throwError . AheadOfLastSync $ Just result
-    pure result
+  query p q ix =
+    withStability ix =<< do
+      let isBefore p' e = p' >= e ^. point
+      let result =
+            mapMaybe (traverse $ predicate q) $
+              ix ^.. events . Lens.folded . Lens.filtered (isBefore p)
+      res <- withStability ix result
+      aheadOfSync <- isAheadOfSync p ix
+      when aheadOfSync $ throwError . AheadOfLastSync $ Just res
+      pure result
```

Comment:
I would avoid introducing `withStability` in generic queries, especially as we can easily opt-in for it if needed.